src/net_processing.cpp

```diff
@@ -505,6 +505,24 @@ using PeerRef = std::shared_ptr<Peer>;
 Mutex g_peer_mutex;
 static std::map<NodeId, PeerRef> g_peer_map GUARDED_BY(g_peer_mutex);
 
+/** Get a reference to each Peer, then call a function on each of them */
+template<typename Callable>
+void ForEachPeer(Callable&& func)
+{
+    std::vector<PeerRef> peer_copies;
+    {
+        LOCK(g_peer_mutex);
+        peer_copies.reserve(g_peer_map.size());
+        for (auto peer : g_peer_map) {
+            peer_copies.push_back(peer.second);
+        }
+    }
+
+    for (auto&& peer : peer_copies) {
+        func(peer);
```

Comment:
575c5f55cc44348f0ebc7740c90013b96bdb8c27
It is good for concurrency to limit the lock scope. But is it safe to call `func` on potentially stale `peer_copies`?

Reply:
Yes. `peer_copies` holds shared_ptr references, so the objects pointed to can't be destructed while it's in scope.

Reply:
I mean not language safety rather client behavior safety: if another thread removes a `peer` from the `g_peer_map`, can we rely on result of `func(peer)`?

Reply:
Yes, because peer still exists in that case.